### PR TITLE
Tensorboard

### LINKED ISSUE #1142: TensorBoard
As most tensorflow users know one of the most powerful features it packs is it's amazing visualization tools. Does keras support tensorboard? Or should we add it in the "implement" list?

### PR DESCRIPTION
**Pull request by tboquet**
Following the discussion in #1142 I have worked on a cleaner callback.

## Features
* works with a `Graph` or a `Sequential` model
* add weights and biases to the log
* add the data flow graph to the log
* add the loss to the log
* if asked, add the accuracy to the log

What could be added easily:
* image summaries
* sparsity measures if the activation is a ReLU

## Notes
We could improve how we pass data to the callbacks but it depends on how Keras will evolve in the future regarding the scopes and the naming of contexts/ops.

For now, the user has to feed the callback with a dictionnary `{tensor: value}` but he has to get it from the `model._test` Keras function. I didn't found a clean way to do this with the same names provided in the `Graph` model or using positions for the `Sequential` model (if there is something obvious I missed le me know :) ).

## Tests
I'm not sure about how to write tests for this feature (and other callbacks), if anyone is willing to guide me on this you're welcome.

Example:

```python
from keras.models import Sequential
from keras.layers.core import Dense, Activation
from keras.callbacks import TensorBoard
import numpy as np
from keras.utils.test_utils import get_test_data
from keras.utils import np_utils

input_dim = 32
nb_hidden = 16
nb_class = 4
batch_size = 32
nb_epoch = 10


def _get_test_data():
    np.random.seed(1234)

    train_samples = 2000
    test_samples = 500

    (X_train, y_train), (X_test, y_test) = get_test_data(nb_train=train_samples,
                                                         nb_test=test_samples,
                                                         input_shape=(input_dim,),
                                                         classification=True,
                                                         nb_class=4)

    y_test = np_utils.to_categorical(y_test)
    y_train = np_utils.to_categorical(y_train)
    return (X_train, y_train), (X_test, y_test)

(X_train, y_train), (X_test, y_test) = _get_test_data()


def test_seq_acc():
    model = Sequential()
    model.add(Dense(nb_hidden, input_shape=(input_dim,)))
    model.add(Activation('relu'))
    model.add(Dense(nb_class))
    model.add(Activation('softmax'))
    model.compile(loss='categorical_crossentropy', optimizer='rmsprop')
    model.summary()

    feed = {model._test.inputs[0]: X_train, model._test.inputs[1]: y_train,
            model._test.inputs[2]: np.ones(len(X_train))}
    tsb = TensorBoard(model=model, feed=feed, log_dir='./logs',
                      show_accuracy=True)

    model.fit(X_train, y_train, batch_size=batch_size,
              nb_epoch=nb_epoch, show_accuracy=True,
              verbose=1, validation_data=(X_test, y_test),
              callbacks=[tsb])

if __name__ == '__main__':
    test_seq_acc()
```

(closes #1142 )

### CODE REVIEW

File: keras/callbacks.py [hunk 1]
```diff
@@ -381,3 +388,69 @@ def __init__(self, schedule):
...
+            if hasattr(cur_layer, 'W'):
+                tf.histogram_summary('{}_W'.format(l), cur_layer.W)
+            if hasattr(cur_layer, 'b'):
+                tf.histogram_summary('{}_b'.format(l), cur_layer.b)
+            if hasattr(cur_layer, 'get_output'):
+                tf.histogram_summary('{}_out'.format(l),
+                                     cur_layer.get_output())
+        if show_accuracy is True and mod_type == 'Sequential':
+            f_output = self.model._test_with_acc
+            tf.scalar_summary(f_output.outputs[1].name,
```
Comment (fchollet) [inline keras/callbacks.py]: Would benefit from being explicitly named "accuracy". Otherwise it turns up as "mean_n" and is very confusing.

File: keras/callbacks.py [hunk 2]
```diff
@@ -381,3 +388,69 @@ def __init__(self, schedule):
...
+                              f_output.outputs[1])
+        else:
+            f_output = self.model._test
+        tf.scalar_summary(f_output.outputs[0].name,
```
Comment (fchollet) [inline keras/callbacks.py]: "loss". In the case of a Graph it can be named explicitly (by the name of the output node); in the case of Sequential just "loss" should be fine. Even better would be to name the type of the loss explicitly, e.g. "binary_crossentropy" (but I'm not sure that's possible).

File: keras/callbacks.py [hunk 3]
```diff
@@ -381,3 +388,69 @@ def __init__(self, schedule):
...
+
+        self.model = model
+        self.freq = freq
+        self.log_dir = log_dir
+        self.sess = tfbe._get_session()
+        self.feed = feed
+
+        mod_type = self.model.get_config()['name']
+        if mod_type == 'Sequential':
+            layers = {l.get_config()['name']: l for l in self.model.layers}
```
Comment (fchollet) [inline keras/callbacks.py]: I think this is a good solution for now. In the future when we have a layer naming system, we'll use that instead.

File: keras/callbacks.py [hunk 4]
```diff
@@ -381,3 +388,69 @@ def __init__(self, schedule):
...
+        log_dir: the path of the directory where to save the log
+            files to be parsed by tensorboard
+    '''
+    def __init__(self, model, feed, freq=2, log_dir='./logs',
+                 show_accuracy=False):
+        super(Callback, self).__init__()
+        assert _BACKEND == 'tensorflow', \
+            'TensorBoard callback only works with the tensorflow backend'
+        import tensorflow as tf
+        import keras.backend.tensorflow_backend as tfbe
```
Comment (fchollet) [inline keras/callbacks.py]: We generally call it "KTF"

File: keras/callbacks.py [hunk 5]
```diff
@@ -381,3 +388,69 @@ def __init__(self, schedule):
 
     def on_epoch_begin(self, epoch, logs={}):
         self.model.optimizer.lr.set_value(self.schedule(epoch))
+
+
+class TensorBoard(Callback):
+    ''' Tensorboard basic visualizations.
+
+    This callback writes a log usable with tensorboard.
```
Comment (fchollet) [inline keras/callbacks.py]: This would benefit from a bit more explanation on how to use TensorBoard, at least a link to https://www.tensorflow.org/versions/master/how_tos/summaries_and_tensorboard/index.html

**Comment by fchollet** [comment]
> I'm not sure about how to write tests for this feature (and other callbacks), if anyone is willing to guide me on this you're welcome.

I believe that merely running a model that has the callback (in a test restricted to TensorFlow) would be enough. The goal would only be to make sure that the code gets run during unit tests, to prevent things breaking in an update down the road. Let me add a `test_callbacks.py` in our unit tests.

File: keras/callbacks.py (hunk 1 above)
Comment (tboquet) [inline keras/callbacks.py]: Thanks to point this out, added!

File: keras/callbacks.py (hunk 2 above)
Comment (tboquet) [inline keras/callbacks.py]: I tried to do my best. When multiple outputs are considered in the `Graph`, we can only access the last op wich is a `add` of all the outputs. I just took all the losses names to name the graph.

**Comment by tboquet** [comment]
I'll let you add the `test_callbacks.py` so you could impose some structure. I have some examples ready to be integrated for this callback. I will have some time next week to add tests for other callbacks too. Thank you for the comments and directions.

File: keras/callbacks.py [hunk 6]
```diff
@@ -381,3 +388,85 @@ def __init__(self, schedule):
...
+                                     cur_layer.get_output())
+        if show_accuracy is True and mod_type == 'Sequential':
+            f_output = self.model._test_with_acc
+            tf.scalar_summary('Accuracy',
+                              f_output.outputs[1])
+            tf.scalar_summary('Loss',
+                              f_output.outputs[0])
+        else:
+            f_output = self.model._test
+            losses = [self.model.loss[loss] for loss in self.model.loss]
```
Comment (fchollet) [inline keras/callbacks.py]: This won't work for a Sequential model. I'd recommend separating the two cases (graph / sequential) entirely, for clarity.

Comment (tboquet) [inline keras/callbacks.py]: Sorry I missed it, I've separated the two cases.

**Comment by fchollet** [comment]
I have added basic tests for existing callbacks. You can add a test for the TensorBoard callback, restricted to TensorFlow.

**Comment by tboquet** [comment]
Thank you! I'll push the tests this morning.

File: keras/callbacks.py [hunk 7]
```diff
@@ -44,9 +45,11 @@ def on_batch_begin(self, batch, logs={}):
             callback.on_batch_begin(batch, logs)
         self._delta_ts_batch_begin.append(time.time() - t_before_callbacks)
         delta_t_median = np.median(self._delta_ts_batch_begin)
-        if self._delta_t_batch > 0. and delta_t_median > 0.95 * self._delta_t_batch and delta_t_median > 0.1:
+        if self._delta_t_batch > 0. and delta_t_median > 0.95 * \
+           self._delta_t_batch and delta_t_median > 0.1:
             warnings.warn('Method on_batch_begin() is slow compared '
-                          'to the batch update (%f). Check your callbacks.' % delta_t_median)
+                          'to the batch update (%f). Check your callbacks.'
+                          % delta_t_median)
```
Comment (fchollet) [inline keras/callbacks.py]: Just wondering -- would the line splitting here work with the string formatting?

Comment (tboquet) [inline keras/callbacks.py]: It seems to work for python but I can change it if it's messy.

**Comment by tboquet** [comment]
Not sure why some tests are not passing since I didn't change other tests or core files.

File: keras/callbacks.py (hunk 7 above)
Comment (fchollet) [inline keras/callbacks.py]: I don't think it's messy, just wanted to make sure.

**Comment by fchollet** [comment]
> Not sure why some tests are not passing since I didn't change other tests or core files.
 Hide all checks


It seems running the TensorBoard test is crashing most subsequent tests. I suspect it's the impact of `KTF._set_session(session)`.

**Comment by tboquet** [comment]
I will try to get the session and restore it after the TensorBoard test.

**Comment by tboquet** [comment]
The TensorFlow part is now working but there is a weird bug about download of the test data with Theano and python 2.7 :(

**Comment by fchollet** [comment]
Looks good to me! You can just squash the commits for this PR into 1-2 compact commits, and I will merge it.

**Comment by tboquet** [comment]
Cool! I have squashed all the past commits. Is it good or do you want me to git reset and rewrite the message to have a cleaner history?

**Comment by fchollet** [comment]
> do you want me to git reset and rewrite the message to have a cleaner history?

Yes, please. The goal is to have the PR fit in 1-2 commits, to maintain history readability.

**Comment by fchollet** [comment]
I realize the whole squashing thing might not be clear, so I've updated CONTRIBUTING.md with a link to http://rebaseandsqua.sh/.

**Comment by tboquet** [comment]
Sorry about that. Thank you for the link, I used the script and I had to merge the changes to the remote.

**Comment by fchollet** [comment]
It seems the commits were not squashed. It's not a big issue, I can do it for you.

While re-reading the code and tests, I realized that there were serious issues with the API, mainly:
- passing a model to the callback constructor. That's not how callbacks should work, the associated model is passed to the callback via the `_set_model()` method.
- redundancy between model `fit` API and callback API, e.g. `show_accuracy`. The user shouldn't have to set this value twice.
- the most important issue: the API is unacceptably complicated. We cannot expect users to pass to the callback a dictionary of the form `feed = {model._test.inputs[0]: X_train, model._test.inputs[1]: y_train, model._test.inputs[2]: np.ones(len(X_train))}`, because:
   - this requires significant knowledge of the Keras internals, which should not be expected from the user
   - this requires significant knowledge of TensorFlow in order to make sense, which should not be expected from the user
   - even for a Keras and TensorFlow expert, it is very cumbersome. All of it should be automated behind the scenes.

Sorry for not noticing these issues earlier, I'm a lousy code reviewer. I need to actually start coding in order to start thinking about what the code does.

I'll take it from there.

**Comment by tboquet** [comment]
No pb, I was also wondering if it was right when trying to integrate both the train and test data to follow the loss and accuracy. It will be a lot easier when the tensorflow backend will integrate naming of the tensors and ops concerned for both Sequential and Graph models.

**Comment by fchollet** [comment]
I made the following changes:

- squashed your commits and rebased with you as the author
- changed the API to `TensorBoard(log_dir=filepath, histogram_freq=0)`
- changed the functionality so that training metrics (accuracy and loss) are displayed as well as evaluation metrics. This is important, for instance to monitor overfitting.

**Comment by tboquet** [comment]
Thank you!

**Comment by fchollet** [comment]
Np! I think this PR was very useful, and it fulfills a lot of the vision of the old "Keras total visualization project" that I never had much time to work on : ) 

Thanks for the valuable contribution!
